File: app/src/main/java/com/qiuminal/zhhhelper/DataLoader.java
package com.qiuminal.zhhhelper;

import android.content.Context;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

/**
 * 从 assets 加载三个码表并合并为内存数据，主键均为「字头」：
 *   zi.txt    字头 + 编码
 *   chai.txt  拆分（两行）+ 拼音 + U码
 *   zheng.txt 整句码
 */
public class DataLoader {

    private static final String FILE_ZI = "zi.txt";
    private static final String FILE_CHAI = "chai.txt";
    private static final String FILE_ZHENG = "zheng.txt";

    private static Map<String, CharData> dataMap;

    /**
     * 加载数据。每次调用都重新从 assets 读取并重建内存表。
     */
    public static synchronized void load(Context context) {
        Map<String, CharData> map = new HashMap<>();

        // 1) zi.txt：字头 + 编码
        try {
            readLines(context, FILE_ZI, line -> {
                int idx = line.indexOf('\t');
                if (idx <= 0) return;
                String key = line.substring(0, idx);
                CharData d = new CharData();
                d.charText = key;
                d.codes = line.substring(idx + 1).trim();
                map.put(key, d);
            });
        } catch (IOException e) {
            e.printStackTrace();
        }

        // 2) chai.txt：拆分两行 + 拼音 + U码
        try {
            readLines(context, FILE_CHAI, line -> {
                String[] parts = line.split("\t", -1);
                if (parts.length < 6) return;
                String key = stripBom(parts[0]).trim();
                if (key.isEmpty()) return;
                CharData d = map.get(key);
                if (d == null) {
                    d = new CharData();
                    d.charText = key;
                    map.put(key, d);
                }
                d.rootCodes = parts[1].trim();            // 拆分第1行
                d.components = parts[2].trim();           // 拆分第2行
                d.pinyin = parts[3].trim();               // 拼音
                d.unicode = concat(parts[4].trim(), parts[5].trim()); // U码（第5+6列拼接）
            });
        } catch (IOException e) {
            e.printStackTrace();
        }

        // 3) zheng.txt：整句码
        try {
            readLines(context, FILE_ZHENG, line -> {
                int idx = line.indexOf('\t');
                if (idx <= 0) return;
                String key = line.substring(0, idx);
                CharData d = map.get(key);
                if (d == null) {
                    d = new CharData();
                    d.charText = key;
                    map.put(key, d);
                }
                d.zhengCode = line.substring(idx + 1).trim();
            });
        } catch (IOException e) {
            e.printStackTrace();
        }

        dataMap = map;
    }

    /**
     * 查询单个字。
     * 输入可能是整句话或带标点的文本，这里自动提取第一个可作为字头的
     * 汉字字符（跳过空白与标点，兼容全角括号等场景）。
     */
    public static CharData query(String text) {
        if (dataMap == null || text == null) return null;
        String key = extractQueryKey(text);
        if (key == null || key.isEmpty()) return null;
        return dataMap.get(key);
    }

    /**
     * 提取第一个非空白、非标点的字符作为查询键。
     */
    private static String extractQueryKey(String text) {
        for (int i = 0; i < text.length(); ) {
            int cp = text.codePointAt(i);
            i += Character.charCount(cp);
            if (Character.isWhitespace(cp)) continue;
            if (!Character.isLetterOrDigit(cp)) continue;
            return new String(Character.toChars(cp));
        }
        return null;
    }

    /**
     * 获取已加载字数
     */
    public static int getCount() {
        return dataMap == null ? 0 : dataMap.size();
    }

    private interface LineHandler {
        void handle(String line);
    }

    private static void readLines(Context context, String asset, LineHandler handler) throws IOException {
        try (InputStream is = context.getAssets().open(asset);
             BufferedReader reader = new BufferedReader(
                     new InputStreamReader(is, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                handler.handle(line);
            }
        }
    }

    private static String stripBom(String s) {
        if (s != null && !s.isEmpty() && s.charAt(0) == '\uFEFF') {
            return s.substring(1);
        }
        return s;
    }

    private static String concat(String a, String b) {
        if (a.isEmpty()) return b;
        if (b.isEmpty()) return a;
        return a + " " + b;
    }
}

File: app/src/main/java/com/qiuminal/zhhhelper/DataLoader.kt
package com.qiuminal.zhhhelper

import android.content.Context
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.zip.Adler32

/**
 * 从 assets/tables.bin 加载码表二进制字典（由 Gradle 任务 generateTablesBin
 * 把 data/ 下三个原始 txt 码表编译而来），按字头 UTF-16 码元二分查找。
 *
 * 二进制格式（大端序，见 app/build.gradle generateTablesBin 注释）：
 *   Header(24B): magic "ZHHT" | version u32 | recordCount u32
 *                | indexOffset u32 | valuesOffset u32 | checksum u32(Adler32)
 *   Index: recordCount * 12B，按 (keyU16_1, keyU16_2) 升序
 *   Values: 每记录 7 字段（codes/rootCodes/components/pinyin/block/code/zheng）
 *
 * 无需构建 HashMap，内存占用小、启动快、数据不可直接编辑。
 */
object DataLoader {

    private const val FILE_BIN = "tables.bin"
    private const val MAGIC = 0x5A484854 // "ZHHT"

    @Volatile
    private var buf: ByteBuffer? = null
    private var count = 0
    private var indexOffset = 0
    private var valuesOffset = 0

    /**
     * 加载二进制码表（幂等，校验 magic/version/checksum，失败则保持空数据）。
     */
    @Synchronized
    fun load(context: Context) {
        try {
            val bytes = context.assets.open(FILE_BIN).use { it.readBytes() }
            val b = ByteBuffer.wrap(bytes) // 默认大端
            if (b.int != MAGIC) return
            val version = b.int
            if (version != 1) return
            count = b.int
            indexOffset = b.int
            valuesOffset = b.int
            val expected = b.int.toLong() and 0xFFFFFFFFL
            val crc = Adler32()
            crc.update(bytes, indexOffset, bytes.size - indexOffset)
            if (crc.value != expected) return
            buf = b
        } catch (e: Exception) {
            // 非关键路径失败：静默降级，避免打扰用户
        }
    }

    fun isLoaded(): Boolean = buf != null

    /**
     * 多字查询：按输入顺序返回每个查到的字的卡片数据。
     * 只跳过空白；是否收录完全以码表为准——
     * 〇(U+3007) 等 Unicode 类别非 Letter/Digit 的汉字此前会被
     * isLetterOrDigit 误过滤导致查不到，故不再按类别过滤。
     */
    fun queryAll(text: String?): List<CharData> {
        val b = buf ?: return emptyList()
        if (text.isNullOrEmpty()) return emptyList()
        val result = ArrayList<CharData>()
        var i = 0
        while (i < text.length) {
            val cp = text.codePointAt(i)
            i += Character.charCount(cp)
            if (Character.isWhitespace(cp)) continue
            val key = String(Character.toChars(cp))
            val d = lookup(b, key) ?: continue
            result.add(d)
        }
        return result
    }

    /** 查询单个字（兼容旧接口）。 */
    fun query(text: String?): CharData? = queryAll(text).firstOrNull()

    /** 已收录字数。 */
    fun getCount(): Int = count

    /** 在索引中二分查找单个字。 */
    private fun lookup(b: ByteBuffer, key: String): CharData? {
        val c1 = key[0].code
        val c2 = if (key.length > 1) key[1].code else 0
        var lo = 0
        var hi = count - 1
        while (lo <= hi) {
            val mid = (lo + hi) ushr 1
            val pos = indexOffset + mid * 12
            val k1 = b.getShort(pos).toInt() and 0xFFFF
            val k2 = b.getShort(pos + 2).toInt() and 0xFFFF
            val cmp = compareKey(c1, k1, c2, k2)
            if (cmp < 0) {
                hi = mid - 1
            } else if (cmp > 0) {
                lo = mid + 1
            } else {
                val off = valuesOffset + b.getInt(pos + 4)
                val len = b.getInt(pos + 8)
                return readRecord(b, off, len, key)
            }
        }
        return null
    }

    /** 按 UTF-16 码元比较两键。 */
    private fun compareKey(a1: Int, b1: Int, a2: Int, b2: Int): Int = when {
        a1 < b1 || (a1 == b1 && a2 < b2) -> -1
        a1 > b1 || (a1 == b1 && a2 > b2) -> 1
        else -> 0
    }

    /** 读取一条记录：7 个长度前缀字段。 */
    private fun readRecord(b: ByteBuffer, off: Int, len: Int, key: String): CharData {
        val d = CharData()
        d.charText = key
        var p = off
        val end = off + len
        var field = 0
        while (p < end && field < 7) {
            val fLen = b.getShort(p).toInt() and 0xFFFF
            p += 2
            val s = String(b.array(), p, fLen, StandardCharsets.UTF_8)
            p += fLen
            when (field) {
                0 -> d.codes = s
                1 -> d.rootCodes = s
                2 -> d.components = s
                3 -> d.pinyin = s
                4 -> d.unicodeBlock = s
                5 -> d.unicodeCode = s
                6 -> d.zhengCode = s
            }
            field++
        }
        return d
    }
}
